### PR TITLE
Create a new array instead of pushing the item

### PR DESCRIPTION
**Pull request by KristofGilis**
TypeError: Cannot add property 1, object is not extensible

Now the array will be recreated

Fixes https://github.com/bithost-gmbh/ngx-mat-select-search/issues/376

### CODE REVIEW

**Comment by macjohnny** [comment]
@KristofGilis thanks for the fix!
i am wondering if that is different from the change we needed to revert in https://github.com/bithost-gmbh/ngx-mat-select-search/pull/388 to avoid the problem described in https://github.com/bithost-gmbh/ngx-mat-select-search/issues/376#issuecomment-1184489849 by @angelaki ?

i think that problem was because in a filtered list of options, only the ones visible can be in the selection, while the filtering is just temporary and the full selection, including of those options that were not among the filtered ones temporarily, should be restored.

**Comment by KristofGilis** [comment]
> @KristofGilis thanks for the fix! i am wondering if that is different from the change we needed to revert in #388 to avoid the problem described in [#376 (comment)](https://github.com/bithost-gmbh/ngx-mat-select-search/issues/376#issuecomment-1184489849) by @angelaki ?
> 
> i think that problem was because in a filtered list of options, only the ones visible can be in the selection, while the filtering is just temporary and the full selection, including of those options that were not among the filtered ones temporarily, should be restored.

It is in typescript always safer to recreate arrays because you never know if it would be extensible or not. I tested in the project where I use the package and it doesn’t throw the error anymore ;)

**Comment by macjohnny** [comment]
@KristofGilis yes, I agree, it would fix https://github.com/bithost-gmbh/ngx-mat-select-search/issues/376, but it would probably re-introduce the problem described in https://github.com/bithost-gmbh/ngx-mat-select-search/issues/376#issuecomment-1184489849

@angelaki can you help testing this?

**Comment by angelaki** [comment]
Sure, I'll check it these days.

**Comment by angelaki** [comment]
Sorry guys, super busy lately. But check it for sure next week!

**Comment by Plekker** [comment]
Any updates on this?

**Comment by angelaki** [comment]
So sorry, got a project with a hard dead-line right now. Haven't forgotten it, will check as soon as possible. Don't want to make wrong estimations again ;) But probably this weekend / start of next week.

**Comment by KristofGilis** [comment]
@angelaki Did you already find some time to check this?

**Comment by angelaki** [comment]
Sorry guys it took like forever to make me finally check this! And yeah, recreating the bound values breaks the regular Angular Select Component's event. You can check in my fork I created for demonstrating the issue: https://github.com/angelaki/ngx-mat-select-search.

Don't have time right now to check the actual reason, but guess it might be pretty hard to solve it. Angular's component it pretty bound to the visual items. If both scenarios are needed, I guess the easiest might be to just introduce a property to switch the behavior?

But @KristofGilis if you find time to investigate this issue any further and find a solution to solve both that would be awesome! Since I'm familiar right now with this code again, feel free to ask if you have any questions :)

**Comment by KristofGilis** [comment]
@angelaki 

I try to understand your demonstration of what is going wrong. Is it correct, you are saying that when you use the selectionChange output and enter a filter for ex. "Bank A". The value of the MatSelectChange event, only contains "Bank A" when you select it and not all the default selected options + the "Bank A" option?

If I run the tester application with your demonstration, I got following values from the MatSelectChange event (what seems correct to me):
image

**Comment by macjohnny** [comment]
@KristofGilis maybe introducing a new option flag as an input would be a good way forward to enable the behavior you are proposing? then people can just choose.

File: src/app/mat-select-search/mat-select-search.component.ts [hunk 1]
```diff
@@ -553,7 +553,7 @@ export class MatSelectSearchComponent implements OnInit, OnDestroy, ControlValue
                 && !optionValues.some(v => this.matSelect.compareWith(v, previousValue))) {
                 // if a value that was selected before is deselected and not found in the options, it was deselected
                 // due to the filtering, so we restore it.
-                values.push(previousValue);
+                values = [...values, previousValue];
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.ts:556]: please do a `values = [...values]` copy before the forEach loop, and then keep the values.push here

**Comment by angelaki** [comment]
Sorry, cannot check the behavior right now / today, but sure will do tomorrow! @macjohnny can you confirm the bug in my branch?

**Comment by macjohnny** [comment]
@angelaki I havent tested your branch, sorry

**Comment by angelaki** [comment]
> @angelaki
> 
> I try to understand your demonstration of what is going wrong. Is it correct, you are saying that when you use the selectionChange output and enter a filter for ex. "Bank A". The value of the MatSelectChange event, only contains "Bank A" when you select it and not all the default selected options + the "Bank A" option?
> 
> If I run the tester application with your demonstration, I got following values from the MatSelectChange event (what seems correct to me): image

Yeah, just checked it again, in my case only *Bank A* gets selected: 
image

1. Open the dropdown (you see all the banks with the initial ones selected)
2. Filter e.g. for "Bank A" (so only "Bank A" is visible!)
3. Select "Bank A"

- The databound property will contain the initial banks + the new selected one
- The event's property will contain the selected one and only those banks, the filter still made visible (so in the case Bank A only)

**Comment by macjohnny** [comment]
@KristofGilis can you add an `@Input()` option to enable this behavior if wanted?

**Comment by KristofGilis** [comment]
> @KristofGilis can you add an `@Input()` option to enable this behavior if wanted?

I added the `@Input()` option

File: src/app/mat-select-search/mat-select-search.component.ts [hunk 2]
```diff
@@ -193,6 +193,11 @@ export class MatSelectSearchComponent implements OnInit, OnDestroy, ControlValue
    */
   @Input() alwaysRestoreSelectedOptionsMulti = false;
 
+  /**
+   * Recreate array of selected values
```
Comment (macjohnny) [inline src/app/mat-select-search/mat-select-search.component.ts:197]:
```suggestion
   * Recreate array of selected values for multi-selects.
   *
   * This is useful if the selected values are stored in an immutable data structure.
```

please also add this to https://github.com/bithost-gmbh/ngx-mat-select-search/blob/master/README.md#inputs

**Review by macjohnny — APPROVED** [review]
thanks for your contribution!

File: README.md [hunk 3]
```diff
@@ -499,6 +499,14 @@ Furthermore, it provides the following inputs:
    * Defaults to false, so selected options are only restored while filtering is active.
    */
   @Input() alwaysRestoreSelectedOptionsMulti = false;
+
+    /**
```
Comment (macjohnny) [inline README.md:503]:
```suggestion
  /**
```

**Pull request merged**